US VISA APPROVAL PROJECT EDA 


LIFECYCLE OF AN ML PROJECT

UNDERSTANDING THE PROBLEM STATEMENT

DATA COLLECTION

EDA 

DATA CLEANING 

DATA PREPROCESSING

MODEL TRAINING 

CHOOSING BEST MODEL

UNDERSTANDING THE PROBLEM STATEMENT AND PLANNING THE NITTY GRITTY AS MUCH AS POSSIBLE.

It means we must thouroughly understand the business problem and the concern that needs to be addressed. 

we must jot down the goals of the experiments and the pain points that we will encounter while addressing and  developing the project.

we must clearly understand the implications of the problem and also the solutions to the problems which we will address.

We must define SMART (specific, measurable, achievable,relevant and time bound) objectives which clearly aligns with our success criterions and also we should not forget to involve the stakeholders at every step of these projects to gain more clarity about the success or the goals to be attained.

We must adhere to the time boundedness and the cost constraints and define our solutions within these realms.

We must allocate resources keeping in mind about the scope of the expansion and scalability of the solutions and also plan the mitigations about the various problems which can arise. 




COLLECTION OF DATA.

To collect the business problem oriented data effectively we must conduct surveys , conduct thourough market research and collect datathrough API'S ,collect data through web scraping and also substantiate the data needed from other avavilable sources. But we must not forget to also chech the quality of the data collected whether it suits our busines problem or not.

Here in this problem statement we already have the data in the form of a csv file.

In [3]:
#lets load the data in pandas dataframe to enhance the features of pandas library

import pandas as pd
df = pd.read_csv('EasyVisa.csv')

UNDERSTANDING THE DATA

In [4]:
df.head() 
#.head method shows us the first five rows of our dataset and by this we also come to know about the structure overview of the data

case_id continent education_of_employee has_job_experience  \
0  EZYV01      Asia           High School                  N   
1  EZYV02      Asia              Master's                  Y   
2  EZYV03      Asia            Bachelor's                  N   
3  EZYV04      Asia            Bachelor's                  N   
4  EZYV05    Africa              Master's                  Y   

  requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                     N            14513         2007                 West   
1                     N             2412         2002            Northeast   
2                     Y            44444         2008                 West   
3                     N               98         1897                 West   
4                     N             1082         2005                South   

   prevailing_wage unit_of_wage full_time_position case_status  
0         592.2029         Hour                  Y      Denied  
1       83425.6500         Year                  Y   Certified  
2      122996.8600         Year                  Y      Denied  
3       83434.0300         Year                  Y      Denied  
4      149907.3900         Year                  Y   Certified

In [5]:
df.shape #by this .shape method we come to know how big is the data by inspecting the number of rows and columns

# we can see that the data has 25480 rows and 12 columns

(25480, 12)

In [8]:
'''sometimes there is some bias or consistency noticed in the data while we do .head() method
so to see some random rows of the dataset we use use .sample() method to reduce the first view bias from our minds '''

df.sample(10) #sample of any 8 rows 

case_id      continent education_of_employee has_job_experience  \
2770    EZYV2771         Europe            Bachelor's                  N   
17302  EZYV17303           Asia           High School                  Y   
1734    EZYV1735           Asia           High School                  N   
2955    EZYV2956           Asia            Bachelor's                  N   
18807  EZYV18808           Asia            Bachelor's                  Y   
16026  EZYV16027           Asia             Doctorate                  Y   
24143  EZYV24144           Asia              Master's                  Y   
3914    EZYV3915  North America            Bachelor's                  Y   
2105    EZYV2106         Europe            Bachelor's                  Y   
5186    EZYV5187           Asia              Master's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
2770                      N             4798         1998   
17302                     N             2123         1896   
1734                      N             1250         1919   
2955                      Y             2532         1985   
18807                     N             2431         2008   
16026                     N             1978         1947   
24143                     N             5834         2003   
3914                      N             2164         2004   
2105                      N             5582         1996   
5186                      N             3412         1976   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
2770                 South       39063.6600         Year                  Y   
17302                South         259.2239         Hour                  Y   
1734                 South       74515.6600         Year                  N   
2955             Northeast       44346.7900         Year                  Y   
18807                 West       33587.1700         Year                  Y   
16026                South         460.9642         Hour                  Y   
24143                 West       82273.6400         Year                  Y   
3914             Northeast       92268.3500         Year                  Y   
2105                 South       92444.6900         Year                  Y   
5186                 South       73482.2400         Year                  Y   

      case_status  
2770    Certified  
17302      Denied  
1734    Certified  
2955       Denied  
18807   Certified  
16026   Certified  
24143   Certified  
3914    Certified  
2105    Certified  
5186    Certified

From the preliminary view we can see that 
1.much of the applications are from ASIA 
2.maximum applicants are having graduate degree
3.most of the applicants have job experience
4.most of the applicants donot require job training
5.most of the applicants' employers are based in south part of USA
6.most of the applicants have a yearly package promised to them i.e. the majority can be of organised sector employees
7.most of the folks have full time offers
8. many have got their visa approved and there is strong correlation of approval with the sector they work in , their education qualifications.
9. if the wage is high then approval status is likely to be approved as we can assume that they are going to have some prestigious jobs in some organised sectors with avg. employees in the comapany ranging from 2500 to 3400 

these are the preliminary overview we get and assumptions we can make.

In [11]:
''' now we use the .info() method to get the datatype of each columns/features, 
to observe the non null values count in each column'''

df.info()

'''we can see that this dataset contains 3 columns with the numerical datatypes(2 int and 1 float) 
and 9 columns with object /string datatypes.
we also observe that this dataset occupies greater than 2.3 mb space in memmory

we can reduce the memory of the dataset by converting the float datatypes to int64 if suitable
this optimisation can be very useful for large datatsets and for ml models to be fitted'''

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25480 entries, 0 to 25479
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   case_id                25480 non-null  object 
 1   continent              25480 non-null  object 
 2   education_of_employee  25480 non-null  object 
 3   has_job_experience     25480 non-null  object 
 4   requires_job_training  25480 non-null  object 
 5   no_of_employees        25480 non-null  int64  
 6   yr_of_estab            25480 non-null  int64  
 7   region_of_employment   25480 non-null  object 
 8   prevailing_wage        25480 non-null  float64
 9   unit_of_wage           25480 non-null  object 
 10  full_time_position     25480 non-null  object 
 11  case_status            25480 non-null  object 
dtypes: float64(1), int64(2), object(9)
memory usage: 2.3+ MB


'we can see that this dataset contains 3 columns with the numerical datatypes(2 int and 1 float) \nand 9 columns with object /string datatypes.\nwe also observe that this dataset occupies greater than 2.3 mb space in memmory\n\nwe can reduce the memory of the dataset by converting the float datatypes to int64 if suitable\nthis optimisation can be very useful for large datatsets and for ml models to be fitted'

In [13]:
# for checking if there are any missing values or what is the total number of missing values in each feature

df.isnull().sum()


case_id                  0
continent                0
education_of_employee    0
has_job_experience       0
requires_job_training    0
no_of_employees          0
yr_of_estab              0
region_of_employment     0
prevailing_wage          0
unit_of_wage             0
full_time_position       0
case_status              0
dtype: int64

we can see that there are no missing values in the dataset

In [14]:
# now  we run the .describe() method to check the basic statistical numbers related to each numerical column which are not object datatype features

df.describe() 

no_of_employees   yr_of_estab  prevailing_wage
count     25480.000000  25480.000000     25480.000000
mean       5667.043210   1979.409929     74455.814592
std       22877.928848     42.366929     52815.942327
min         -26.000000   1800.000000         2.136700
25%        1022.000000   1976.000000     34015.480000
50%        2109.000000   1997.000000     70308.210000
75%        3504.000000   2005.000000    107735.512500
max      602069.000000   2016.000000    319210.270000

We can see the various statistics of the numerical columns like mean ,standard deviation, 1st,2nd,3rd quartile values , min max values 

We can see that in number of employees in a company which is the minimum value is -26 which is not possible. Suspicions must arise and this company must be evaluated for correct informations.

The minimum prevailing wage is also 2.1367 which is also not feasible so the apllicant must be scrutinised thouroughly. 

this types of outliers are rare and extreme which must be imputed or removed if possible else this points can seriously harm the model we want to fit and affect the results badly.

In [15]:
# now we check how many rows are duplicated in the dataset

df.duplicated().sum()

# there are 0 duplicate rows as per the findings 

0

In [19]:
# suppose there were duplicate rows then we could have used the 
df.drop_duplicates()

#as there as 0 duplicate so it returns the original dataset with 0 rows removed

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
2          Denied  
3          Denied  
4       Certified  
...           ...  
25475   Certified  
25476   Certified  
25477   Certified  
25478   Certified  
25479   Certified  

[25480 rows x 12 columns]